File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/BingoRoomSpot.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkSpot;
import systems.zlink.framework.spots.ZLinkSpotActorJoinResult;
import systems.zlink.framework.spots.ZLinkSpotContext;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.framework.spots.ZLinkSpotRelocationReadyCompletion;
import systems.zlink.framework.spots.ZLinkTimer;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.server.play.domain.bingo.BingoGame;
import systems.zlink.samples.bingo.server.play.domain.bingo.BingoRoomGame;
import systems.zlink.samples.bingo.server.play.domain.bingo.BingoRoomModels;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers.BingoRoomSettingsInitializer;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers.BingoRoomTimerHandler;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

public final class BingoRoomSpot implements ZLinkSpot<PlayerActor> {
    private static final Logger logger = LoggerFactory.getLogger(BingoRoomSpot.class);

    private final ZLinkSpotContext context;
    private final BingoRoomSettingsInitializer settingsInitializer;
    private final Map<String, PlayerActor> actors = new HashMap<>();
    private final Map<String, PlayerActor> observers = new HashMap<>();
    private final Map<String, Messages.BingoRoomJoinReq> pendingJoins = new HashMap<>();
    private final List<Messages.BingoRewardAcquiredEvent> pendingObserverRewards =
            new ArrayList<>();
    private BingoRoomModels.BingoRoomSettings settings =
            BingoRoomModels.BingoRoomSettings.create(
                    "two-player", 0, SampleTimings.DrawPeriod.toMillis());
    private BingoRoomGame game;
    private ZLinkTimer timer;
    private boolean cleanupStarted;

    public BingoRoomSpot(
            ZLinkSpotContext context, BingoRoomSettingsInitializer settingsInitializer) {
        this.context = context;
        this.settingsInitializer = settingsInitializer;
        this.game = BingoGame.room(context.spotId(), settings);
    }

    @Override
    public ZLinkSpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkSpotCreateResponse> onCreate(ZLinkMessage request) {
        return CompletableFuture.completedFuture(settingsInitializer.handle(this, request));
    }

    @Override
    public CompletionStage<ZLinkSpotActorJoinResult> onActorJoin(
            String actorId, ZLinkMessage request) {
        Messages.BingoRoomJoinReq joinRequest = request.decode(Messages.BingoRoomJoinReq.class);
        validateJoin(actorId, joinRequest);
        Messages.BingoRoomState preview =
                joinRequest.getObserveOnly()
                        ? observerJoinState(joinRequest)
                        : game.previewJoin(actorId, joinRequest.getDisplayName());
        pendingJoins.put(actorId, joinRequest);
        return CompletableFuture.completedFuture(
                ZLinkSpotActorJoinResult.accept(BingoMessages.bingoRoomJoinRes(preview)));
    }

    @Override
    public CompletionStage<Void> onJoinedActor(PlayerActor actor) {
        Messages.BingoRoomJoinReq request = pendingJoins.get(actor.actorId());
        if (request == null) {
            throw new IllegalStateException("joined actor does not have a pending admission");
        }
        if (request.getObserveOnly()) {
            pendingJoins.remove(actor.actorId());
            join(actor, request, 0, 0);
            List<Messages.BingoRewardAcquiredEvent> pendingRewards =
                    List.copyOf(pendingObserverRewards);
            pendingObserverRewards.clear();
            if (pendingRewards.isEmpty()) {
                return CompletableFuture.completedFuture(null);
            }
            return CompletableFuture.allOf(
                    pendingRewards.stream()
                            .map(this::notifyObservers)
                            .map(CompletionStage::toCompletableFuture)
                            .toArray(CompletableFuture[]::new));
        }
        // --8<-- [start:doc-bingo-room-join]
        return context.outbound()
                .requestToChannel(
                        SampleNames.ApiChannel, BingoMessages.getPlayerRecordReq(actor.actorId()))
                .timeout(SampleTimings.RequestTimeout)
                .yield(Messages.GetPlayerRecordRes.class)
                .thenCompose(
                        record -> {
                            if (pendingJoins.get(actor.actorId()) != request
                                    || game == null
                                    || !game.canAcceptPlayer()) {
                                pendingJoins.remove(actor.actorId());
                                return context.leaveActor(actor);
                            }
                            pendingJoins.remove(actor.actorId());
                            join(actor, request, record.getWins(), record.getLosses());
                            logger.info(
                                    "bingo-record fetched actor={} wins={} losses={}",
                                    actor.actorId(),
                                    record.getWins(),
                                    record.getLosses());
                            return CompletableFuture.completedFuture((Void) null);
                        });
        // --8<-- [end:doc-bingo-room-join]
    }

    @Override
    public CompletionStage<Void> onLeaveActor(PlayerActor actor) {
        if (!actors.containsKey(actor.actorId()) || game == null) {
            observers.remove(actor.actorId());
            logger.info("bingo-lifecycle room-leave actor={}", actor.actorId());
            return CompletableFuture.completedFuture(null);
        }
        Messages.BingoRoomState finalState = game.snapshot();
        Messages.ReportBingoResultReq report =
                BingoMessages.reportBingoResultReq(
                        finalState.getRoomId(),
                        actor.actorId(),
                        finalState.getWinnersList().contains(actor.actorId()),
                        finalState.getDrawSeq());
        return context.outbound()
                .requestToChannel(SampleNames.ApiChannel, report)
                .timeout(SampleTimings.RequestTimeout)
                .yield(Messages.ReportBingoResultRes.class)
                .thenAccept(
                        record -> {
                            logger.info(
                                    "bingo-record reported actor={} wins={} losses={}",
                                    actor.actorId(),
                                    record.getWins(),
                                    record.getLosses());
                            actors.remove(actor.actorId());
                            logger.info("bingo-lifecycle room-leave actor={}", actor.actorId());
                        });
    }

    @Override
    public CompletionStage<Void> onDisconnectActor(PlayerActor actor) {
        actor.markDisconnected();
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onInitialize() {
        if (settings.observerMode()) {
            return CompletableFuture.completedFuture(null);
        }
        return context.addTimer(
                        "bingo-draw",
                        Duration.ofMillis(settings.drawPeriodMillis()),
                        BingoRoomTimerHandler.class,
                        null)
                .thenAccept(created -> timer = created);
    }

    @Override
    public CompletionStage<Void> onClosing() {
        return timer == null ? CompletableFuture.completedFuture(null) : timer.cancel();
    }

    @Override
    public CompletionStage<Void> onRelocationReadyCompleted(
            ZLinkSpotRelocationReadyCompletion completion) {
        return CompletableFuture.completedFuture(null);
    }

    public Messages.BingoRoomJoinRes join(
            PlayerActor actor, Messages.BingoRoomJoinReq request, int wins, int losses) {
        validateJoin(actor.actorId(), request);
        actor.setDisplayName(request.getDisplayName());
        actor.joinRoom(request.getRoomId());
        if (request.getObserveOnly()) {
            observers.put(actor.actorId(), actor);
            context.relocationReady().defer();
            return BingoMessages.bingoRoomJoinRes(observerJoinState(request));
        }
        actors.put(actor.actorId(), actor);
        BingoRoomGame.Change change =
                game.join(actor.actorId(), request.getDisplayName(), wins, losses);
        publishEvents(change.events(), actors::get);
        return BingoMessages.bingoRoomJoinRes(change.state());
    }

    private void validateJoin(String actorId, Messages.BingoRoomJoinReq request) {
        if (!actorId.equals(request.getActorId())) {
            throw new IllegalStateException("Join request actor id does not match bound actor.");
        }
        if (!request.getObserveOnly() && !request.getRoomId().equals(context.spotId())) {
            throw new IllegalStateException("Join request room id does not match bingo room.");
        }
        if (request.getObserveOnly()) {
            if (!settings.observerMode()
                    || !request.getRoomId().equals(settings.observedRoomId())) {
                throw new IllegalStateException(
                        "Observe-only actor can join only its observer BingoRoom.");
            }
            return;
        }
        if (settings.observerMode()) {
            throw new IllegalStateException("Player actor cannot join an observer BingoRoom.");
        }
    }

    public Messages.SubmitBingoCardRes submitCard(
            PlayerActor actor, Messages.SubmitBingoCardReq request) {
        if (!request.getRoomId().equals(context.spotId())) {
            throw new IllegalStateException("Submit request room id does not match bingo room.");
        }
        if (game == null) {
            throw new IllegalStateException("Observer BingoRoom does not own game state.");
        }
        BingoRoomGame.Change change = game.submitCard(actor.actorId(), request.getCardList());
        publishEvents(change.events(), actors::get);
        return BingoMessages.submitBingoCardRes(change.state());
    }

    // --8<-- [start:doc-bingo-draw-timer]
    public CompletionStage<Void> tick() {
        if (game == null || cleanupStarted) {
            return CompletableFuture.completedFuture(null);
        }
        BingoRoomGame.Change change = game.drawNext();
        publishEvents(change.events(), actors::get);
        return publishWinner(change)
                .thenCompose(ignored -> leaveFinishedActors(change))
                .thenRun(
                        () -> {
                            // --8<-- [start:doc-relocation-ready]
                            if (change.state().getStatus().equals("Finished")) {
                                context.relocationReady().defer();
                            }
                            // --8<-- [end:doc-relocation-ready]
                        });
    }

    // --8<-- [end:doc-bingo-draw-timer]

    public CompletionStage<Void> announceReward(Messages.BingoRewardAcquiredEvent event) {
        if (!settings.observerMode() || !event.getRoomId().equals(settings.observedRoomId())) {
            return CompletableFuture.completedFuture(null);
        }
        if (observers.isEmpty()) {
            pendingObserverRewards.add(event);
            return CompletableFuture.completedFuture(null);
        }
        return notifyObservers(event).thenRun(() -> context.relocationReady().defer());
    }

    private CompletionStage<Void> notifyObservers(Messages.BingoRewardAcquiredEvent event) {
        List<CompletionStage<Void>> notifications = new ArrayList<>();
        for (PlayerActor observer : List.copyOf(observers.values())) {
            notifications.add(observer.push(rewardNotification(event)));
        }
        return CompletableFuture.allOf(
                notifications.stream()
                        .map(CompletionStage::toCompletableFuture)
                        .toArray(CompletableFuture[]::new));
    }

    private Messages.BingoRewardAnnouncedNotify rewardNotification(
            Messages.BingoRewardAcquiredEvent event) {
        return BingoMessages.bingoRewardAnnouncedNotify(
                event.getRoomId(),
                event.getActorId(),
                event.getDrawSeq(),
                event.getItemId(),
                event.getItemName(),
                event.getRarity());
    }

    public Messages.StopObservingBingoEventsRes stopObserving(
            PlayerActor actor, Messages.StopObservingBingoEventsReq request) {
        if (!settings.observerMode()
                || !request.getRoomId().equals(settings.observedRoomId())
                || !observers.containsKey(actor.actorId())) {
            return BingoMessages.stopObservingBingoEventsRes(false);
        }
        observers.remove(actor.actorId());
        context.leaveActor(actor).exceptionally(error -> null);
        return BingoMessages.stopObservingBingoEventsRes(true);
    }

    public void applySettings(BingoRoomModels.BingoRoomSettings settings) {
        if (!settings.observerMode() && settings.requiredPlayers() <= 0) {
            throw new IllegalStateException("Bingo room requires at least one player.");
        }
        if (settings.maxDrawNumber() <= 0) {
            throw new IllegalStateException("Bingo room requires at least one draw number.");
        }
        if (settings.drawPeriodMillis() <= 0) {
            throw new IllegalStateException("Bingo room draw period must be positive.");
        }
        this.settings = settings;
        this.game = settings.observerMode() ? null : BingoGame.room(context.spotId(), settings);
        this.cleanupStarted = false;
    }

    RelocationState captureRelocationState() {
        Messages.BingoRoomState state =
                game == null
                        ? BingoMessages.bingoRoomState(
                                context.spotId(),
                                "Running",
                                "",
                                false,
                                0,
                                null,
                                List.of(),
                                List.of(),
                                List.of())
                        : game.snapshot();
        return new RelocationState(settings, state);
    }

    void restoreRelocationState(RelocationState state) {
        settings = state.settings();
        game =
                settings.observerMode()
                        ? null
                        : BingoRoomGame.restore(context.spotId(), settings, state.state());
        cleanupStarted = false;
    }

    record RelocationState(
            BingoRoomModels.BingoRoomSettings settings, Messages.BingoRoomState state) {}

    // --8<-- [start:doc-bingo-room-cleanup]
    private CompletionStage<Void> leaveFinishedActors(BingoRoomGame.Change change) {
        if (cleanupStarted || !change.state().getStatus().equals("Finished")) {
            return CompletableFuture.completedFuture(null);
        }

        cleanupStarted = true;
        List<CompletableFuture<Void>> leaves = new ArrayList<>();
        for (PlayerActor actor : actors.values().toArray(PlayerActor[]::new)) {
            actor.markForDestroyAfterRoomLeave();
            leaves.add(context.leaveActor(actor).toCompletableFuture());
        }
        return CompletableFuture.allOf(leaves.toArray(CompletableFuture[]::new));
    }

    // --8<-- [end:doc-bingo-room-cleanup]

    private CompletionStage<Void> publishWinner(BingoRoomGame.Change change) {
        if (!change.state().getStatus().equals("Finished")
                || change.state().getWinnersList().isEmpty()) {
            return CompletableFuture.completedFuture(null);
        }
        String winner = change.state().getWinnersList().getFirst();
        // --8<-- [start:doc-bingo-reward-publish]
        return context.outbound()
                .publish(
                        SampleNames.RoomRewardChannel,
                        SampleNames.WinnerTopic,
                        BingoMessages.bingoRewardAcquiredEvent(
                                change.state().getRoomId(),
                                winner,
                                change.state().getDrawSeq(),
                                "rare-golden-dauber",
                                "Golden Dauber",
                                "Legendary"))
                .submit();
        // --8<-- [end:doc-bingo-reward-publish]
    }

    private void publishEvents(
            List<BingoRoomModels.RoomEvent> events, Function<String, PlayerActor> actorResolver) {
        for (BingoRoomModels.RoomEvent event : events) {
            publishEvent(event, actorResolver.apply(event.recipientActorId()));
        }
    }

    private void publishEvent(BingoRoomModels.RoomEvent event, PlayerActor recipient) {
        if (recipient == null) {
            return;
        }
        switch (event.kind()) {
            case PLAYER_JOINED ->
                    recipient.push(
                            BingoMessages.playerJoinedNotify(
                                    event.state().getRoomId(),
                                    event.joinedActorId(),
                                    event.joinedDisplayName(),
                                    event.seat(),
                                    event.host(),
                                    event.state()));
            case GAME_STARTED ->
                    recipient.push(BingoMessages.bingoGameStartedNotify(event.state()));
            case NUMBER_DRAWN ->
                    recipient.push(
                            BingoMessages.bingoNumberDrawnNotify(
                                    event.state().getRoomId(),
                                    event.state().getDrawSeq(),
                                    event.drawnNumber(),
                                    event.state()));
            case STATE -> recipient.push(BingoMessages.bingoStateNotify(event.state()));
            case GAME_ENDED -> recipient.push(BingoMessages.bingoGameEndedNotify(event.state()));
        }
    }

    private Messages.BingoRoomState observerJoinState(Messages.BingoRoomJoinReq request) {
        return BingoMessages.bingoRoomState(
                request.getRoomId(),
                "Running",
                "",
                false,
                0,
                null,
                List.of(),
                List.of(),
                List.of());
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/BingoRoomSpot.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot

import java.time.Duration
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.ZLinkSuspendingSpot
import systems.zlink.framework.kotlin.yieldReply
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkSpotActorJoinResult
import systems.zlink.framework.spots.ZLinkSpotClosingContext
import systems.zlink.framework.spots.ZLinkSpotContext
import systems.zlink.framework.spots.ZLinkSpotCreateResponse
import systems.zlink.framework.spots.ZLinkSpotRelocationReadyCompletion
import systems.zlink.framework.spots.ZLinkTimer
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoGame
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomEvent
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomEventKind
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomGame
import systems.zlink.samples.kotlin.bingo.server.play.domain.bingo.BingoRoomSettings
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers.BingoRoomSettingsInitializer
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers.BingoRoomTimerHandler
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameEndedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameStartedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoNumberDrawnNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRewardAcquiredEvent
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRewardAnnouncedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomState
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoStateNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.GetPlayerRecordReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.GetPlayerRecordRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.PlayerJoinedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReportBingoResultReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReportBingoResultRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.card
import systems.zlink.samples.kotlin.bingo.shared.contracts.winners

class BingoRoomSpot(
    override val context: ZLinkSpotContext,
    private val settingsInitializer: BingoRoomSettingsInitializer,
) : ZLinkSuspendingSpot<PlayerActor>() {
    private val logger = LoggerFactory.getLogger(BingoRoomSpot::class.java)

    private val actors = mutableMapOf<String, PlayerActor>()
    private val observers = mutableMapOf<String, PlayerActor>()
    private val pendingJoins = mutableMapOf<String, BingoRoomJoinReq>()
    private val pendingObserverRewards = mutableListOf<BingoRewardAcquiredEvent>()
    private var settings =
        BingoRoomSettings.create("two-player", 0, SampleTimings.DrawPeriod.toMillis())
    private var game: BingoRoomGame? = BingoGame.room(context.spotId(), settings)
    private var timer: ZLinkTimer? = null
    private var cleanupStarted = false

    override suspend fun onCreateSuspending(request: ZLinkMessage): ZLinkSpotCreateResponse {
        settingsInitializer.handle(this, request)
        return ZLinkSpotCreateResponse.accept()
    }

    override suspend fun onActorJoinSuspending(
        actorId: String,
        request: ZLinkMessage,
    ): ZLinkSpotActorJoinResult {
        val joinRequest = request.decode(BingoRoomJoinReq::class.java)
        validateJoin(actorId, joinRequest)
        val preview =
            if (joinRequest.observeOnly) {
                observerJoinState(joinRequest)
            } else {
                requireGame().previewJoin(actorId, joinRequest.displayName)
            }
        pendingJoins[actorId] = joinRequest
        return ZLinkSpotActorJoinResult.accept(BingoRoomJoinRes(preview))
    }

    override suspend fun onJoinedActorSuspending(actor: PlayerActor) {
        val request =
            pendingJoins[actor.actorId()] ?: error("joined actor does not have a pending admission")
        if (request.observeOnly) {
            pendingJoins.remove(actor.actorId())
            join(actor, request, 0, 0)
            val pendingRewards = pendingObserverRewards.toList()
            pendingObserverRewards.clear()
            pendingRewards.forEach { notifyObservers(it) }
            return
        }
        // --8<-- [start:doc-bingo-room-join]
        val record =
            context
                .outbound()
                .requestToChannel(SampleNames.ApiChannel, GetPlayerRecordReq(actor.actorId()))
                .timeout(SampleTimings.RequestTimeout)
                .yieldReply<GetPlayerRecordRes>()
        val game = this.game
        if (pendingJoins[actor.actorId()] !== request || game == null || !game.canAcceptPlayer()) {
            pendingJoins.remove(actor.actorId())
            context.leaveActor(actor).await()
            return
        }
        pendingJoins.remove(actor.actorId())
        join(actor, request, record.wins, record.losses)
        logger.info(
            "bingo-record fetched actor={} wins={} losses={}",
            actor.actorId(),
            record.wins,
            record.losses,
        )
        // --8<-- [end:doc-bingo-room-join]
    }

    override suspend fun onLeaveActorSuspending(actor: PlayerActor) {
        if (!actors.containsKey(actor.actorId()) || game == null) {
            observers.remove(actor.actorId())
            logger.info("bingo-lifecycle room-leave actor={}", actor.actorId())
            return
        }
        val state = requireGame().snapshot()
        val record =
            context
                .outbound()
                .requestToChannel(
                    SampleNames.ApiChannel,
                    ReportBingoResultReq(
                        state.roomId,
                        actor.actorId(),
                        state.winners.contains(actor.actorId()),
                        state.drawSeq,
                    ),
                )
                .timeout(SampleTimings.RequestTimeout)
                .yieldReply<ReportBingoResultRes>()
        logger.info(
            "bingo-record reported actor={} wins={} losses={}",
            actor.actorId(),
            record.wins,
            record.losses,
        )
        actors.remove(actor.actorId())
        logger.info("bingo-lifecycle room-leave actor={}", actor.actorId())
    }

    override suspend fun onDisconnectActorSuspending(actor: PlayerActor) {
        actor.markDisconnected()
    }

    override suspend fun onInitializeSuspending() {
        if (settings.observerMode()) {
            return
        }
        timer =
            context
                .addTimer(
                    "bingo-draw",
                    Duration.ofMillis(settings.drawPeriodMillis),
                    BingoRoomTimerHandler::class.java,
                    null,
                )
                .await()
    }

    override suspend fun onClosingSuspending(context: ZLinkSpotClosingContext) {
        timer?.cancel()?.await()
    }

    override suspend fun onRelocationReadyCompletedSuspending(
        completion: ZLinkSpotRelocationReadyCompletion
    ) = Unit

    fun join(
        actor: PlayerActor,
        request: BingoRoomJoinReq,
        wins: Int,
        losses: Int,
    ): BingoRoomJoinRes {
        validateJoin(actor.actorId(), request)
        actor.setDisplayName(request.displayName)
        actor.joinRoom(request.roomId)
        if (request.observeOnly) {
            observers[actor.actorId()] = actor
            context.relocationReady().defer()
            return BingoRoomJoinRes(observerJoinState(request))
        }
        val change = requireGame().join(actor.actorId(), request.displayName, wins, losses)
        actors[actor.actorId()] = actor
        publishEvents(change.events, actors::get)
        return BingoRoomJoinRes(change.state)
    }

    private fun validateJoin(actorId: String, request: BingoRoomJoinReq) {
        if (request.actorId != actorId) {
            throw IllegalStateException("Join request actor id does not match bound actor.")
        }
        if (!request.observeOnly && request.roomId != context.spotId()) {
            throw IllegalStateException("Join request room id does not match bingo room.")
        }
        if (request.observeOnly) {
            if (!settings.observerMode() || request.roomId != settings.observedRoomId) {
                throw IllegalStateException(
                    "Observe-only actor can join only its observer BingoRoom."
                )
            }
            return
        }
        if (settings.observerMode()) {
            throw IllegalStateException("Player actor cannot join an observer BingoRoom.")
        }
    }

    suspend fun submitCard(actor: PlayerActor, request: SubmitBingoCardReq): SubmitBingoCardRes {
        if (request.roomId != context.spotId()) {
            throw IllegalStateException("Submit request room id does not match bingo room.")
        }
        val change = requireGame().submitCard(actor.actorId(), request.card)
        publishEvents(change.events, actors::get)
        return SubmitBingoCardRes(change.state)
    }

    // --8<-- [start:doc-bingo-draw-timer]
    suspend fun tick() {
        val game = this.game
        if (game == null || cleanupStarted) {
            return
        }
        val change = game.drawNext()
        publishEvents(change.events, actors::get)
        publishWinner(change)
        leaveFinishedActors(change)
        if (change.state.status == BingoRoomGame.Finished) {
            context.relocationReady().defer()
        }
    }

    // --8<-- [end:doc-bingo-draw-timer]

    suspend fun announceReward(event: BingoRewardAcquiredEvent) {
        if (!settings.observerMode() || event.roomId != settings.observedRoomId) {
            return
        }
        if (observers.isEmpty()) {
            pendingObserverRewards += event
            return
        }
        notifyObservers(event)
        context.relocationReady().defer()
    }

    private suspend fun notifyObservers(event: BingoRewardAcquiredEvent) {
        for (observer in observers.values.toList()) {
            observer
                .push(
                    BingoRewardAnnouncedNotify(
                        event.roomId,
                        event.actorId,
                        event.drawSeq,
                        event.itemId,
                        event.itemName,
                        event.rarity,
                    )
                )
                .await()
        }
    }

    suspend fun stopObserving(
        actor: PlayerActor,
        request: StopObservingBingoEventsReq,
    ): StopObservingBingoEventsRes {
        if (
            !settings.observerMode() ||
                request.roomId != settings.observedRoomId ||
                !observers.containsKey(actor.actorId())
        ) {
            return StopObservingBingoEventsRes(false)
        }
        observers.remove(actor.actorId())
        context.leaveActor(actor).exceptionally { null }
        return StopObservingBingoEventsRes(true)
    }

    // --8<-- [start:doc-bingo-room-cleanup]
    private suspend fun leaveFinishedActors(change: BingoRoomGame.Change) {
        if (cleanupStarted || change.state.status != BingoRoomGame.Finished) {
            return
        }
        cleanupStarted = true
        for (actor in actors.values.toList()) {
            actor.markForDestroyAfterRoomLeave()
            context.leaveActor(actor).await()
        }
    }

    // --8<-- [end:doc-bingo-room-cleanup]

    fun applySettings(settings: BingoRoomSettings) {
        check(settings.observerMode() || settings.requiredPlayers > 0) {
            "Bingo room requires at least one player."
        }
        check(settings.maxDrawNumber > 0) { "Bingo room requires at least one draw number." }
        check(settings.drawPeriodMillis > 0) { "Bingo room draw period must be positive." }
        this.settings = settings
        game = if (settings.observerMode()) null else BingoGame.room(context.spotId(), settings)
        cleanupStarted = false
    }

    internal fun captureRelocationState(): RelocationState =
        RelocationState(
            settings,
            game?.snapshot()
                ?: BingoRoomState(
                    context.spotId(),
                    BingoRoomGame.Running,
                    "",
                    false,
                    0,
                    null,
                    emptyList(),
                    emptyList(),
                    emptyList(),
                ),
        )

    internal fun restoreRelocationState(state: RelocationState) {
        settings = state.settings
        game =
            if (settings.observerMode()) {
                null
            } else {
                BingoRoomGame.restore(context.spotId(), settings, state.state)
            }
        cleanupStarted = false
    }

    internal data class RelocationState(val settings: BingoRoomSettings, val state: BingoRoomState)

    private suspend fun publishWinner(change: BingoRoomGame.Change) {
        val state = change.state
        if (state.status != BingoRoomGame.Finished || state.winners.isEmpty()) {
            return
        }
        val winner = state.winners.first()
        // --8<-- [start:doc-bingo-reward-publish]
        context
            .outbound()
            .publish(
                SampleNames.RoomRewardChannel,
                SampleNames.WinnerTopic,
                BingoRewardAcquiredEvent(
                    state.roomId,
                    winner,
                    state.drawSeq,
                    "rare-golden-dauber",
                    "Golden Dauber",
                    "Legendary",
                ),
            )
            .submit()
            .await()
        // --8<-- [end:doc-bingo-reward-publish]
    }

    private fun publishEvents(
        events: List<BingoRoomEvent>,
        actorResolver: (String) -> PlayerActor?,
    ) {
        for (event in events) {
            publishEvent(event, actorResolver(event.recipientActorId))
        }
    }

    private fun publishEvent(event: BingoRoomEvent, recipient: PlayerActor?) {
        if (recipient == null) {
            return
        }
        when (event.kind) {
            BingoRoomEventKind.PLAYER_JOINED ->
                recipient.push(
                    PlayerJoinedNotify(
                        event.state.roomId,
                        event.joinedActorId!!,
                        event.joinedDisplayName!!,
                        event.seat,
                        event.host,
                        event.state,
                    )
                )

            BingoRoomEventKind.GAME_STARTED -> recipient.push(BingoGameStartedNotify(event.state))

            // --8<-- [start:doc-bingo-bound-push]
            BingoRoomEventKind.NUMBER_DRAWN ->
                recipient.push(
                    BingoNumberDrawnNotify(
                        event.state.roomId,
                        event.state.drawSeq,
                        event.drawnNumber,
                        event.state,
                    )
                )
            // --8<-- [end:doc-bingo-bound-push]

            BingoRoomEventKind.STATE -> recipient.push(BingoStateNotify(event.state))

            BingoRoomEventKind.GAME_ENDED -> recipient.push(BingoGameEndedNotify(event.state))
        }
    }

    private fun observerJoinState(request: BingoRoomJoinReq): BingoRoomState =
        BingoRoomState(
            request.roomId,
            BingoRoomGame.Running,
            "",
            false,
            0,
            null,
            emptyList(),
            emptyList(),
            emptyList(),
        )

    private fun requireGame(): BingoRoomGame =
        game ?: throw IllegalStateException("Observer BingoRoom does not own game state.")
}
